# <center><u>**`Data Analysis using Pandas🐼`**</u></center>

## <center><u>**Black Friday Sales Prediction**</u></center>

`Pandas` is an open source library for data analysis in Python. It is one of those packages that makes analysing data much easier. It helps to manage data and it used 2 data types for storing data:
- series 
- dataframe.

Dataframe like an excel spreadsheet that is storing data. A dataframe is made up of several series. Each column of a dataframe is a series.


`Series` :- It is one dimensional,labeled and homogeneous array.

`DataFrames` :- It is two dimensional tabular data structure.

## <u>Black-Friday-Dataset</u>
This dataset comprises of sales transactions captured at a retail store. It’s a classic dataset to explore and expand your feature engineering skills and day to day understanding from multiple shopping experiences.The dataset has 550,069 rows and 12 columns.

Problem: Predict purchase amount.

### Task 1. Importing Pandas

In [103]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

### Task 2. Change default values for displaying rows and columns.

In [104]:
# display default maximum rows
pd.get_option("display.max_rows")

60

In [105]:
# display default maximum columns
pd.get_option("display.max_columns")

20


### Task 3. Read dataset using pandas.

In [106]:
# As we have data in CSV format so lets use read_csv for loading dataset
df=pd.read_csv("C:\\Users\\sejal\\Downloads\\blackfriday.csv")
df

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      NaN                 NaN      8370  
1                      6.0                14.0     15200  
2                      NaN                 NaN      1422  
3                     14.0                 NaN      1057  
4                      NaN                 NaN      7969  
...                    ...                 ...       ...  
550063                 NaN                 NaN       368  
550064                 NaN                 NaN       371  
550065                 NaN                 NaN       137  
550066                 NaN                 NaN       365  
550067                 NaN                 NaN       490  

[550068 rows x 12 columns]

In [107]:
# Lets check type of df
type(df)

pandas.core.frame.DataFrame

### Task 4. Find shape of dataframe i.e number of rows and columns.

In [108]:
# Now lets check the shape of dataframe.
df.shape

(550068, 12)

- No. of Rows : 550068
- No. of Columns : 12

In [109]:
# Check dimension of dataframe
df.ndim 

2

## <u>Exploring the dataset</u>

### Task 5. Print top 10 rows and last 10 rows from dataset

In [110]:
# Lets view the first five rows of the dataframe using head()
df.head(10)

User_ID Product_ID Gender    Age  Occupation City_Category  \
0  1000001  P00069042      F   0-17          10             A   
1  1000001  P00248942      F   0-17          10             A   
2  1000001  P00087842      F   0-17          10             A   
3  1000001  P00085442      F   0-17          10             A   
4  1000002  P00285442      M    55+          16             C   
5  1000003  P00193542      M  26-35          15             A   
6  1000004  P00184942      M  46-50           7             B   
7  1000004  P00346142      M  46-50           7             B   
8  1000004   P0097242      M  46-50           7             B   
9  1000005  P00274942      M  26-35          20             A   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   
5                          3               0                   1   
6                          2               1                   1   
7                          2               1                   1   
8                          2               1                   1   
9                          1               1                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN      1422  
3                14.0                 NaN      1057  
4                 NaN                 NaN      7969  
5                 2.0                 NaN     15227  
6                 8.0                17.0     19215  
7                15.0                 NaN     15854  
8                16.0                 NaN     15686  
9                 NaN                 NaN      7871

In [111]:
# We can also check rows from last using tail().By default , it also give last 5. Lets check for last 10 rows
df.tail(10)

User_ID Product_ID Gender    Age  Occupation City_Category  \
550058  1006024  P00372445      M  26-35          12             A   
550059  1006025  P00370853      F  26-35           1             B   
550060  1006026  P00371644      M  36-45           6             C   
550061  1006029  P00372445      F  26-35           1             C   
550062  1006032  P00372445      M  46-50           7             A   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
550058                          0               1                  20   
550059                          1               0                  19   
550060                          1               1                  20   
550061                          1               1                  20   
550062                          3               0                  20   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
550058                 NaN                 NaN       121  
550059                 NaN                 NaN        48  
550060                 NaN                 NaN       494  
550061                 NaN                 NaN       599  
550062                 NaN                 NaN       473  
550063                 NaN                 NaN       368  
550064                 NaN                 NaN       371  
550065                 NaN                 NaN       137  
550066                 NaN                 NaN       365  
550067                 NaN                 NaN       490

### Task 6. Check if there is any duplicate rows in dataset.

In [112]:
#checking for duplicate rows
sum(df.duplicated())

0

### Task 7. Create new dataset with all product related columns.

In [113]:
# Lets check columns present in dataframe
df.columns

Index(['User_ID', 'Product_ID', 'Gender', 'Age', 'Occupation', 'City_Category',
       'Stay_In_Current_City_Years', 'Marital_Status', 'Product_Category_1',
       'Product_Category_2', 'Product_Category_3', 'Purchase'],
      dtype='object')

In [114]:
# here we need the columns which contain product related detail
[i for i in df.columns if 'Product' in i]

['Product_ID',
 'Product_Category_1',
 'Product_Category_2',
 'Product_Category_3']

In [115]:
# Lets create new dataframe with these product related columns
#Subset of dataframe
df_product=df[['Product_ID','Product_Category_1','Product_Category_2','Product_Category_3']]
df_product

Product_ID  Product_Category_1  Product_Category_2  Product_Category_3
0       P00069042                   3                 NaN                 NaN
1       P00248942                   1                 6.0                14.0
2       P00087842                  12                 NaN                 NaN
3       P00085442                  12                14.0                 NaN
4       P00285442                   8                 NaN                 NaN
...           ...                 ...                 ...                 ...
550063  P00372445                  20                 NaN                 NaN
550064  P00375436                  20                 NaN                 NaN
550065  P00375436                  20                 NaN                 NaN
550066  P00375436                  20                 NaN                 NaN
550067  P00371644                  20                 NaN                 NaN

[550068 rows x 4 columns]

In [116]:
# returns a list of the row axis labels as well as columns
df.axes

[RangeIndex(start=0, stop=550068, step=1),
 Index(['User_ID', 'Product_ID', 'Gender', 'Age', 'Occupation', 'City_Category',
        'Stay_In_Current_City_Years', 'Marital_Status', 'Product_Category_1',
        'Product_Category_2', 'Product_Category_3', 'Purchase'],
       dtype='object')]

In [117]:
# values returns the Series of values as ndarray.
df.Gender.values

array(['F', 'F', 'F', ..., 'F', 'F', 'F'], dtype=object)

In [118]:
df.Age.values

array(['0-17', '0-17', '0-17', ..., '26-35', '55+', '46-50'], dtype=object)

In [119]:
#  Size returns the number of elements in the underlying dataset/column. Lets try with dataset first.
df.size #Columns X Rows

6600816

In [120]:
df.Gender.size

550068

### Task 8. Check for all columns datatype and related counts

In [121]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  object 
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  object 
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Product_Category_3          166821 non-null  float64
 11  Purchase                    550068 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 50.4+ MB


In [122]:
# Lets only check datatype of each columns
df.dtypes

User_ID                         int64
Product_ID                     object
Gender                         object
Age                            object
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years     object
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
Purchase                        int64
dtype: object

### Change datatype of 'Purchase' to float.

In [123]:
# Lets try to change 'Purchase' to float datatype
df["Purchase"]=df["Purchase"].astype("Float64")

# lets check
df.dtypes  # Purchse = float64

User_ID                         int64
Product_ID                     object
Gender                         object
Age                            object
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years     object
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
Purchase                      Float64
dtype: object

In [124]:
# change it back to int
df.dtypes

User_ID                         int64
Product_ID                     object
Gender                         object
Age                            object
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years     object
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
Purchase                      Float64
dtype: object

### Task 9. Generate descriptive statistical values for numerical columns

The `describe` method shows basic statistical characteristics of each numerical feature (int64 and float64 types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles

In [125]:
# Lets use describe on top of our dataset
df.describe()

User_ID     Occupation  Marital_Status  Product_Category_1  \
count  5.500680e+05  550068.000000   550068.000000       550068.000000   
mean   1.003029e+06       8.076707        0.409653            5.404270   
std    1.727592e+03       6.522660        0.491770            3.936211   
min    1.000001e+06       0.000000        0.000000            1.000000   
25%    1.001516e+06       2.000000        0.000000            1.000000   
50%    1.003077e+06       7.000000        0.000000            5.000000   
75%    1.004478e+06      14.000000        1.000000            8.000000   
max    1.006040e+06      20.000000        1.000000           20.000000   

       Product_Category_2  Product_Category_3     Purchase  
count       376430.000000       166821.000000     550068.0  
mean             9.842329           12.668243  9263.968713  
std              5.086590            4.125338  5023.065394  
min              2.000000            3.000000         12.0  
25%              5.000000            9.000000       5823.0  
50%              9.000000           14.000000       8047.0  
75%             15.000000           16.000000      12054.0  
max             18.000000           18.000000      23961.0

### Task 10. Try to use same function on categorical columns

We can also use describe for other datatype using include options

In [126]:
# Apart from int and float , we have datatype as object . Lets do for those columns
df.describe(include=["object"])

Product_ID  Gender     Age City_Category Stay_In_Current_City_Years
count      550068  550068  550068        550068                     550068
unique       3631       2       7             3                          5
top     P00265242       M   26-35             B                          1
freq         1880  414259  219587        231173                     193821

### Task 11. Get percentage distribution of each product id available in dataset and find with highest occuring product id in dataset.

    We can use value_counts for checking different values in categorical
    columns

In [127]:
# lets check for Product_ID
df["Product_ID"].value_counts()
# product id -- purchased counts

P00265242    1880
P00025442    1615
P00110742    1612
P00112142    1562
P00057642    1470
             ... 
P00314842       1
P00298842       1
P00231642       1
P00204442       1
P00066342       1
Name: Product_ID, Length: 3631, dtype: int64

In [128]:
# Get normalized value 
df["Product_ID"].value_counts(normalize=True)

P00265242    0.003418
P00025442    0.002936
P00110742    0.002931
P00112142    0.002840
P00057642    0.002672
               ...   
P00314842    0.000002
P00298842    0.000002
P00231642    0.000002
P00204442    0.000002
P00066342    0.000002
Name: Product_ID, Length: 3631, dtype: float64

In [129]:
# Multiply it with 100 to get percentage value and round it up to 3 decimal place
round(df["Product_ID"].value_counts(normalize=True)*100,3)

P00265242    0.342
P00025442    0.294
P00110742    0.293
P00112142    0.284
P00057642    0.267
             ...  
P00314842    0.000
P00298842    0.000
P00231642    0.000
P00204442    0.000
P00066342    0.000
Name: Product_ID, Length: 3631, dtype: float64

In [130]:
# To check sum of all percentage values
round(df["Product_ID"].value_counts(normalize=True)*100,3).sum()

99.97600000000001

## <u>Handling Missing values</u>

### Task 12 . Check for columns having null values and count of null contaning rows.

Pandas provides isna() and notna() functions to detect 'NA' values.

detect ‘NA’ values in the dataframe : `df.isna().sum()`

detect ‘NA’ values in a particular column in the dataframe : `pd.isna(df[‘col_name’])` , `df[‘col_name’].notna()`

`isnull()` is just an alias of the `isna()` method in pandas source code.

In [131]:
# Lets check on our dataset 
df.isnull().sum()

User_ID                            0
Product_ID                         0
Gender                             0
Age                                0
Occupation                         0
City_Category                      0
Stay_In_Current_City_Years         0
Marital_Status                     0
Product_Category_1                 0
Product_Category_2            173638
Product_Category_3            383247
Purchase                           0
dtype: int64

We can see 2 columns i.e. `Product_Category_2` and `Product_Category_2` with missing values and count is 173638 and 383247 respectively.

In [132]:
# Print Product_Category_2
df.Product_Category_2

0          NaN
1          6.0
2          NaN
3         14.0
4          NaN
          ... 
550063     NaN
550064     NaN
550065     NaN
550066     NaN
550067     NaN
Name: Product_Category_2, Length: 550068, dtype: float64

In [133]:
# Print Product_Category_3
df.Product_Category_3

0          NaN
1         14.0
2          NaN
3          NaN
4          NaN
          ... 
550063     NaN
550064     NaN
550065     NaN
550066     NaN
550067     NaN
Name: Product_Category_3, Length: 550068, dtype: float64

### Task 13. Handle missing value using dropping and imputing both options.

There are several methods to handle missing values. Each method has its own advantages and disadvantages.

1. Dropping
2. Imputting (Mean / Mode / Median / ffill / bfill)

In [134]:
# Lets first try with drop option. Dropping Product_Category_2 using axis 1

#In pandas axis = 0 refers to horizontal axis 
#or rows and axis = 1 refers to vertical axis or columns.
df_temp=df.drop("Product_Category_3",axis=1,inplace=False)

df_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 11 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  object 
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  object 
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Purchase                    550068 non-null  Float64
dtypes: Float64(1), float64(1), int64(4), object(5)
memory usage: 46.7+ MB


In [135]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  object 
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  object 
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Product_Category_3          166821 non-null  float64
 11  Purchase                    550068 non-null  Float64
dtypes: Float64(1), float64(2), int64(4), object(5)
memory usage: 50.9+ MB


In [136]:
# Get index for all rows with Product_Category_3 missing
df_temp.Product_Category_2.isnull()

0          True
1         False
2          True
3         False
4          True
          ...  
550063     True
550064     True
550065     True
550066     True
550067     True
Name: Product_Category_2, Length: 550068, dtype: bool

In [137]:
# Get index for all rows with Product_Category_2 missing
df_temp[df_temp.Product_Category_2.isnull()].index

Int64Index([     0,      2,      4,      9,     11,     12,     21,     22,
                23,     26,
            ...
            550058, 550059, 550060, 550061, 550062, 550063, 550064, 550065,
            550066, 550067],
           dtype='int64', length=173638)

In [138]:
# Drop Product_Category_3 using axis 0
df_temp=df_temp.drop(df_temp[df_temp.Product_Category_2.isnull()].index,axis=0 , inplace=False)

In [139]:
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 376430 entries, 1 to 545914
Data columns (total 11 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     376430 non-null  int64  
 1   Product_ID                  376430 non-null  object 
 2   Gender                      376430 non-null  object 
 3   Age                         376430 non-null  object 
 4   Occupation                  376430 non-null  int64  
 5   City_Category               376430 non-null  object 
 6   Stay_In_Current_City_Years  376430 non-null  object 
 7   Marital_Status              376430 non-null  int64  
 8   Product_Category_1          376430 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Purchase                    376430 non-null  Float64
dtypes: Float64(1), float64(1), int64(4), object(5)
memory usage: 34.8+ MB


`method="pad"`
- method="pad" (ya "ffill") means "forward fill"

In [140]:
df = df.fillna(method="ffill")

In [141]:
df.isnull().sum()

User_ID                       0
Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            1
Product_Category_3            1
Purchase                      0
dtype: int64

In [142]:
# Impute using backword filling for 1 missing value
df = df.fillna(method="bfill")

In [143]:
df.isnull().sum()

User_ID                       0
Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            0
Product_Category_3            0
Purchase                      0
dtype: int64

## <u>Indexing and slicing in pandas </u>

### Task 14. Print age and occupation columns using loc and select 1st, 5th and 10th rows with 1st, 4th and 7th columns using iloc

Here now we will see how to slice and dice the data and get the subset of pandas dataframe.

In [144]:
df1 = df.copy()
df1

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      6.0                14.0    8370.0  
1                      6.0                14.0   15200.0  
2                      6.0                14.0    1422.0  
3                     14.0                14.0    1057.0  
4                     14.0                14.0    7969.0  
...                    ...                 ...       ...  
550063                 2.0                11.0     368.0  
550064                 2.0                11.0     371.0  
550065                 2.0                11.0     137.0  
550066                 2.0                11.0     365.0  
550067                 2.0                11.0     490.0  

[550068 rows x 12 columns]

In [145]:
# select first row of dataframe using loc
df1.loc[0,'Product_ID']

'P00069042'

In [146]:
# Lets try to print 'Purchase' for all rows using loc
df1.loc[:,"Purchase"]

0          8370.0
1         15200.0
2          1422.0
3          1057.0
4          7969.0
           ...   
550063      368.0
550064      371.0
550065      137.0
550066      365.0
550067      490.0
Name: Purchase, Length: 550068, dtype: Float64

In [147]:
#select first five rows for a specific column 'purchase'
df1.loc[:4,"Purchase"]

0     8370.0
1    15200.0
2     1422.0
3     1057.0
4     7969.0
Name: Purchase, dtype: Float64

In [148]:
# Print 'Age' and 'Occupation' using loc
df1.loc[:,['Age','Occupation']]

Age  Occupation
0        0-17          10
1        0-17          10
2        0-17          10
3        0-17          10
4         55+          16
...       ...         ...
550063  51-55          13
550064  26-35           1
550065  26-35          15
550066    55+           1
550067  46-50           0

[550068 rows x 2 columns]

In [149]:
# We can also specify row label. 
# Lets print first five rows with Age and Occupation columns
df1.loc[:4,['Age','Occupation']]

Age  Occupation
0  0-17          10
1  0-17          10
2  0-17          10
3  0-17          10
4   55+          16

In [150]:
# Same result we can verify using head()
df1.loc[:,['Age','Occupation']].head()

Age  Occupation
0  0-17          10
1  0-17          10
2  0-17          10
3  0-17          10
4   55+          16

In [151]:
# Lets try to print all columns for 2nd , 3rd and 4th rows
df1.loc[2:4,:]

User_ID Product_ID Gender   Age  Occupation City_Category  \
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
2                 6.0                14.0    1422.0  
3                14.0                14.0    1057.0  
4                14.0                14.0    7969.0

#### Integer position based indexing using .iloc indexer

In [152]:
# print first row using iloc
df1.iloc[0]

User_ID                         1000001
Product_ID                    P00069042
Gender                                F
Age                                0-17
Occupation                           10
City_Category                         A
Stay_In_Current_City_Years            2
Marital_Status                        0
Product_Category_1                    3
Product_Category_2                  6.0
Product_Category_3                 14.0
Purchase                         8370.0
Name: 0, dtype: object

In [153]:
#select last row of dataframe using iloc
df1.iloc[-1]

User_ID                         1006039
Product_ID                    P00371644
Gender                                F
Age                               46-50
Occupation                            0
City_Category                         B
Stay_In_Current_City_Years           4+
Marital_Status                        1
Product_Category_1                   20
Product_Category_2                  2.0
Product_Category_3                 11.0
Purchase                          490.0
Name: 550067, dtype: object

In [154]:
# select first five columns of data frame with all rows using iloc
df1.iloc[:,:5]

User_ID Product_ID Gender    Age  Occupation
0       1000001  P00069042      F   0-17          10
1       1000001  P00248942      F   0-17          10
2       1000001  P00087842      F   0-17          10
3       1000001  P00085442      F   0-17          10
4       1000002  P00285442      M    55+          16
...         ...        ...    ...    ...         ...
550063  1006033  P00372445      M  51-55          13
550064  1006035  P00375436      F  26-35           1
550065  1006036  P00375436      F  26-35          15
550066  1006038  P00375436      F    55+           1
550067  1006039  P00371644      F  46-50           0

[550068 rows x 5 columns]

In [155]:
# select 1st, 5th and 10th rows with 1st, 4th and 7th columns using iloc
df1.iloc[[0,4,9],[0,3,6]]

User_ID    Age Stay_In_Current_City_Years
0  1000001   0-17                          2
4  1000002    55+                         4+
9  1000005  26-35                          1

### Task 15 . fetch row having maximum purchase amount with complete row details

Pandas provide two functions `idxmax()` and `idxmin()` that return index of first occurrence of maximum or minimum values over requested axis. NA/null values are excluded from the output.

In [156]:
# get index of first occurence of maximum Purchase value 
max_purchase_index = df1['Purchase'].idxmax()
max_purchase_index

87440

In [157]:
min_purchase_index = df1['Purchase'].idxmin()
min_purchase_index

545946

In [158]:
# get the value of the maximum purchase amount
max_purchase_value = df1.Purchase[df1['Purchase'].idxmax()]
max_purchase_value

23961.0

In [159]:
# get the row with the maximum Purchase value using loc
df1[df1['Purchase']==23961]

User_ID Product_ID Gender    Age  Occupation City_Category  \
87440   1001474  P00052842      M  26-35           4             A   
93016   1002272  P00052842      M  26-35           0             C   
370891  1003160  P00052842      M  26-35          17             C   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
87440                           2               1                  10   
93016                           1               0                  10   
370891                          3               0                  10   

        Product_Category_2  Product_Category_3  Purchase  
87440                 15.0                 8.0   23961.0  
93016                 15.0                 8.0   23961.0  
370891                15.0                 8.0   23961.0

So there is three users with maximum amount of purchase : `23961`

### Task 16. Get the purchase amount from 3rd row

- `at` is used to access a value by label `(index and column name)`.
- `iat` is used to access a value by integer position `(row and column number)`.

In [160]:
df1.iat[3,11]

1057.0

In [161]:
df

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      6.0                14.0    8370.0  
1                      6.0                14.0   15200.0  
2                      6.0                14.0    1422.0  
3                     14.0                14.0    1057.0  
4                     14.0                14.0    7969.0  
...                    ...                 ...       ...  
550063                 2.0                11.0     368.0  
550064                 2.0                11.0     371.0  
550065                 2.0                11.0     137.0  
550066                 2.0                11.0     365.0  
550067                 2.0                11.0     490.0  

[550068 rows x 12 columns]

In [162]:
# get value at 3rd row and 11th column pair
df1.iat[2,11]

1422.0

### Task 16. find the purchase amount for a user_id (1006039) and product_id (P00371644)

We can also use Boolean indexing to filter and select the data

- `|` for or
- `&` for and
- `~` for not

In [163]:
# get the purchase amount with a given user_id and product_id
df1.loc[((df1['User_ID']==1006039) & (df1['Product_ID']=='P00371644')),'Purchase']

550067    490.0
Name: Purchase, dtype: Float64

### Task 18. find the user those are in city 'A' with more than 4 years and purchase amount more than 10000

In [164]:
# get the purchase amount with a given user_id and product_id
df1[(df1['City_Category']=='A')&(df1['Stay_In_Current_City_Years']=='4+')&(df1['Purchase']>1000)]

User_ID Product_ID Gender    Age  Occupation City_Category  \
98      1000022  P00351142      M  18-25          15             A   
99      1000022  P00213242      M  18-25          15             A   
100     1000022  P00195942      M  18-25          15             A   
101     1000022  P00115642      M  18-25          15             A   
102     1000022   P0098242      M  18-25          15             A   
...         ...        ...    ...    ...         ...           ...   
545794  1006019  P00070342      M  26-35           0             A   
545795  1006019  P00016842      M  26-35           0             A   
545796  1006019  P00159242      M  26-35           0             A   
545831  1006028  P00178042      M  18-25           4             A   
545832  1006028   P0097242      M  18-25           4             A   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
98                             4+               0                   1   
99                             4+               0                   5   
100                            4+               0                   3   
101                            4+               0                   8   
102                            4+               0                   1   
...                           ...             ...                 ...   
545794                         4+               0                   1   
545795                         4+               0                   1   
545796                         4+               0                   1   
545831                         4+               0                   8   
545832                         4+               0                   1   

        Product_Category_2  Product_Category_3  Purchase  
98                     8.0                17.0   12099.0  
99                     8.0                17.0    8797.0  
100                    4.0                17.0   10681.0  
101                   14.0                17.0    7801.0  
102                   16.0                17.0   15517.0  
...                    ...                 ...       ...  
545794                 2.0                14.0   15577.0  
545795                 2.0                15.0    7617.0  
545796                 2.0                16.0    8299.0  
545831                14.0                16.0    7939.0  
545832                16.0                16.0   19682.0  

[21880 rows x 12 columns]

### Task 19. Discard all female users those are in city 'B' with 3 years and purchase amount less than 5000

In [165]:
df1[~((df1['Gender']=='F') & (df1['City_Category']=='B') & (df1['Stay_In_Current_City_Years']==3) & (df1['Purchase']<5000))]

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      6.0                14.0    8370.0  
1                      6.0                14.0   15200.0  
2                      6.0                14.0    1422.0  
3                     14.0                14.0    1057.0  
4                     14.0                14.0    7969.0  
...                    ...                 ...       ...  
550063                 2.0                11.0     368.0  
550064                 2.0                11.0     371.0  
550065                 2.0                11.0     137.0  
550066                 2.0                11.0     365.0  
550067                 2.0                11.0     490.0  

[550068 rows x 12 columns]

### Task 20. find the record in dataset with below details 

`[1006038 , 'P00375436','F','55+', 1 ,'C','2', 0 , 20, 2.0 , 11.0 , 365 ]`

DataFrame also has an `isin()` method. When calling isin, we pass a set of values as either an array or dict. If values is an array, isin returns a DataFrame of booleans that is the same shape as the original DataFrame, with True wherever the element is in the sequence of values.

In [166]:
# Lets use isin function for searching row with given values
values = [1006038,'P00375436','F','55+',1,'C','2',0,20,2.0,11.0,365]
df1_indexed=df1.isin(values)
df1_indexed

User_ID  Product_ID  Gender    Age  Occupation  City_Category  \
0         False       False    True  False       False          False   
1         False       False    True  False       False          False   
2         False       False    True  False       False          False   
3         False       False    True  False       False          False   
4         False       False   False   True       False           True   
...         ...         ...     ...    ...         ...            ...   
550063    False       False   False  False       False          False   
550064    False        True    True  False        True           True   
550065    False        True    True  False       False          False   
550066     True        True    True   True        True           True   
550067    False       False    True  False        True          False   

        Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                             True            True               False   
1                             True            True                True   
2                             True            True               False   
3                             True            True               False   
4                            False            True               False   
...                            ...             ...                 ...   
550063                       False            True                True   
550064                       False            True                True   
550065                       False            True                True   
550066                        True            True                True   
550067                       False            True                True   

        Product_Category_2  Product_Category_3  Purchase  
0                    False               False     False  
1                    False               False     False  
2                    False               False     False  
3                    False               False     False  
4                    False               False     False  
...                    ...                 ...       ...  
550063                True                True     False  
550064                True                True     False  
550065                True                True     False  
550066                True                True      True  
550067                True                True     False  

[550068 rows x 12 columns]

In [167]:
# Lets use all condition
df1_indexed=df1.isin(values).all(axis=1)
df1[df1_indexed]

User_ID Product_ID Gender  Age  Occupation City_Category  \
550066  1006038  P00375436      F  55+           1             C   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
550066                          2               0                  20   

        Product_Category_2  Product_Category_3  Purchase  
550066                 2.0                11.0     365.0

### Task 21. Visualize records with Occupation value 10 and mask everything left

 - what condition to look for
-  specify what the values should be replaced with when the condition is met. By default, `mask` replaces them with `NaN` , (a way to represent missing data).

In [168]:
# lets use mask function to get only rows with Occupation 10
new_df=df1.mask(df1['Occupation'] != 10)
new_df

User_ID Product_ID Gender   Age  Occupation City_Category  \
0       1000001.0  P00069042      F  0-17        10.0             A   
1       1000001.0  P00248942      F  0-17        10.0             A   
2       1000001.0  P00087842      F  0-17        10.0             A   
3       1000001.0  P00085442      F  0-17        10.0             A   
4             NaN        NaN    NaN   NaN         NaN           NaN   
...           ...        ...    ...   ...         ...           ...   
550063        NaN        NaN    NaN   NaN         NaN           NaN   
550064        NaN        NaN    NaN   NaN         NaN           NaN   
550065        NaN        NaN    NaN   NaN         NaN           NaN   
550066        NaN        NaN    NaN   NaN         NaN           NaN   
550067        NaN        NaN    NaN   NaN         NaN           NaN   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2             0.0                 3.0   
1                               2             0.0                 1.0   
2                               2             0.0                12.0   
3                               2             0.0                12.0   
4                             NaN             NaN                 NaN   
...                           ...             ...                 ...   
550063                        NaN             NaN                 NaN   
550064                        NaN             NaN                 NaN   
550065                        NaN             NaN                 NaN   
550066                        NaN             NaN                 NaN   
550067                        NaN             NaN                 NaN   

        Product_Category_2  Product_Category_3  Purchase  
0                      6.0                14.0    8370.0  
1                      6.0                14.0   15200.0  
2                      6.0                14.0    1422.0  
3                     14.0                14.0    1057.0  
4                      NaN                 NaN      <NA>  
...                    ...                 ...       ...  
550063                 NaN                 NaN      <NA>  
550064                 NaN                 NaN      <NA>  
550065                 NaN                 NaN      <NA>  
550066                 NaN                 NaN      <NA>  
550067                 NaN                 NaN      <NA>  

[550068 rows x 12 columns]

## <u>Sorting in Pandas</u>

### Task 22. Sort dataset row wise and column wise

### Sorting by label

The default value of `axis=0`. Here sorting will be done by rows.

If we set `axis=1`, sorting will be done by columns.

In [169]:
# Sort dataset row wise
df1.sort_index() #By default row wise ascending

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
550063  1006033  P00372445      M  51-55          13             B   
550064  1006035  P00375436      F  26-35           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550066  1006038  P00375436      F    55+           1             C   
550067  1006039  P00371644      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      6.0                14.0    8370.0  
1                      6.0                14.0   15200.0  
2                      6.0                14.0    1422.0  
3                     14.0                14.0    1057.0  
4                     14.0                14.0    7969.0  
...                    ...                 ...       ...  
550063                 2.0                11.0     368.0  
550064                 2.0                11.0     371.0  
550065                 2.0                11.0     137.0  
550066                 2.0                11.0     365.0  
550067                 2.0                11.0     490.0  

[550068 rows x 12 columns]

In [170]:
# Sort dataset column wise
df1.sort_index(axis=1)  
#(axis=1 define column wise alphabetically or ascending)

Age City_Category Gender  Marital_Status  Occupation  \
0        0-17             A      F               0          10   
1        0-17             A      F               0          10   
2        0-17             A      F               0          10   
3        0-17             A      F               0          10   
4         55+             C      M               0          16   
...       ...           ...    ...             ...         ...   
550063  51-55             B      M               1          13   
550064  26-35             C      F               0           1   
550065  26-35             B      F               1          15   
550066    55+             C      F               0           1   
550067  46-50             B      F               1           0   

        Product_Category_1  Product_Category_2  Product_Category_3 Product_ID  \
0                        3                 6.0                14.0  P00069042   
1                        1                 6.0                14.0  P00248942   
2                       12                 6.0                14.0  P00087842   
3                       12                14.0                14.0  P00085442   
4                        8                14.0                14.0  P00285442   
...                    ...                 ...                 ...        ...   
550063                  20                 2.0                11.0  P00372445   
550064                  20                 2.0                11.0  P00375436   
550065                  20                 2.0                11.0  P00375436   
550066                  20                 2.0                11.0  P00375436   
550067                  20                 2.0                11.0  P00371644   

        Purchase Stay_In_Current_City_Years  User_ID  
0         8370.0                          2  1000001  
1        15200.0                          2  1000001  
2         1422.0                          2  1000001  
3         1057.0                          2  1000001  
4         7969.0                         4+  1000002  
...          ...                        ...      ...  
550063     368.0                          1  1006033  
550064     371.0                          3  1006035  
550065     137.0                         4+  1006036  
550066     365.0                          2  1006038  
550067     490.0                         4+  1006039  

[550068 rows x 12 columns]

In [171]:
# Sort row wise in descending order
df1.sort_index(ascending=False)

User_ID Product_ID Gender    Age  Occupation City_Category  \
550067  1006039  P00371644      F  46-50           0             B   
550066  1006038  P00375436      F    55+           1             C   
550065  1006036  P00375436      F  26-35          15             B   
550064  1006035  P00375436      F  26-35           1             C   
550063  1006033  P00372445      M  51-55          13             B   
...         ...        ...    ...    ...         ...           ...   
4       1000002  P00285442      M    55+          16             C   
3       1000001  P00085442      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
0       1000001  P00069042      F   0-17          10             A   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
550067                         4+               1                  20   
550066                          2               0                  20   
550065                         4+               1                  20   
550064                          3               0                  20   
550063                          1               1                  20   
...                           ...             ...                 ...   
4                              4+               0                   8   
3                               2               0                  12   
2                               2               0                  12   
1                               2               0                   1   
0                               2               0                   3   

        Product_Category_2  Product_Category_3  Purchase  
550067                 2.0                11.0     490.0  
550066                 2.0                11.0     365.0  
550065                 2.0                11.0     137.0  
550064                 2.0                11.0     371.0  
550063                 2.0                11.0     368.0  
...                    ...                 ...       ...  
4                     14.0                14.0    7969.0  
3                     14.0                14.0    1057.0  
2                      6.0                14.0    1422.0  
1                      6.0                14.0   15200.0  
0                      6.0                14.0    8370.0  

[550068 rows x 12 columns]

### Sorting by values

Sorting Particular Row or Column
Pandas provides `sort_values()` method to sort by values. It accepts a `by` argument which will use the column name of the DataFrame with which the values are to be sorted.

In [172]:
# Lets try to sort dataset using Purchase column
df1.sort_values(by=['Purchase'])

User_ID Product_ID Gender    Age  Occupation City_Category  \
549221  1004806  P00370293      M  26-35          17             C   
549477  1005184  P00370293      M  18-25          20             B   
547819  1002802  P00370853      M  36-45          20             B   
548027  1003105  P00370853      M  36-45          12             C   
547538  1002402  P00370853      M  46-50          17             B   
...         ...        ...    ...    ...         ...           ...   
292083  1003045  P00052842      M  46-50           1             B   
503697  1005596  P00117642      M  36-45          12             B   
370891  1003160  P00052842      M  26-35          17             C   
87440   1001474  P00052842      M  26-35           4             A   
93016   1002272  P00052842      M  26-35           0             C   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
549221                          2               0                  19   
549477                         4+               1                  19   
547819                          2               0                  19   
548027                         4+               1                  19   
547538                         4+               0                  19   
...                           ...             ...                 ...   
292083                          2               1                  10   
503697                          1               0                  10   
370891                          3               0                  10   
87440                           2               1                  10   
93016                           1               0                  10   

        Product_Category_2  Product_Category_3  Purchase  
549221                 2.0                11.0      12.0  
549477                 2.0                11.0      12.0  
547819                 2.0                11.0      12.0  
548027                 2.0                11.0      12.0  
547538                 2.0                11.0      12.0  
...                    ...                 ...       ...  
292083                15.0                17.0   23960.0  
503697                16.0                 9.0   23960.0  
370891                15.0                 8.0   23961.0  
87440                 15.0                 8.0   23961.0  
93016                 15.0                 8.0   23961.0  

[550068 rows x 12 columns]

In [173]:
# Sort by multiple columns
df1.sort_values(by=['Age','City_Category'])

User_ID Product_ID Gender   Age  Occupation City_Category  \
0       1000001  P00069042      F  0-17          10             A   
1       1000001  P00248942      F  0-17          10             A   
2       1000001  P00087842      F  0-17          10             A   
3       1000001  P00085442      F  0-17          10             A   
85      1000019  P00112542      M  0-17          10             A   
...         ...        ...    ...   ...         ...           ...   
549874  1005766  P00375436      M   55+          14             C   
549925  1005834  P00371644      M   55+          16             C   
549989  1005922  P00370853      M   55+           3             C   
550030  1005980  P00372445      M   55+           1             C   
550066  1006038  P00375436      F   55+           1             C   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
85                              3               0                   1   
...                           ...             ...                 ...   
549874                          0               1                  20   
549925                         4+               0                  20   
549989                          3               0                  19   
550030                          3               0                  20   
550066                          2               0                  20   

        Product_Category_2  Product_Category_3  Purchase  
0                      6.0                14.0    8370.0  
1                      6.0                14.0   15200.0  
2                      6.0                14.0    1422.0  
3                     14.0                14.0    1057.0  
85                    11.0                15.0    7746.0  
...                    ...                 ...       ...  
549874                 2.0                11.0     611.0  
549925                 2.0                11.0     121.0  
549989                 2.0                11.0      12.0  
550030                 2.0                11.0     376.0  
550066                 2.0                11.0     365.0  

[550068 rows x 12 columns]

In [174]:
# Sort in descending order
df1.sort_values(by='Age',ascending=False)

User_ID Product_ID Gender   Age  Occupation City_Category  \
364776  1002092  P00015642      M   55+           1             B   
295862  1003618  P00154042      M   55+          17             A   
295864  1003618  P00004242      M   55+          17             A   
295865  1003618  P00320342      M   55+          17             A   
295866  1003618  P00138542      M   55+          17             A   
...         ...        ...    ...   ...         ...           ...   
231838  1005757  P00145442      M  0-17          10             C   
231839  1005757  P00127642      M  0-17          10             C   
231840  1005757  P00119642      M  0-17          10             C   
231942  1005768  P00173442      F  0-17           7             C   
0       1000001  P00069042      F  0-17          10             A   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
364776                         4+               1                   8   
295862                         4+               1                   1   
295864                         4+               1                   1   
295865                         4+               1                   1   
295866                         4+               1                   5   
...                           ...             ...                 ...   
231838                          2               0                   8   
231839                          2               0                   1   
231840                          2               0                   1   
231942                          3               0                   8   
0                               2               0                   3   

        Product_Category_2  Product_Category_3  Purchase  
364776                16.0                17.0    5998.0  
295862                 2.0                15.0   11390.0  
295864                 6.0                 8.0   12086.0  
295865                 8.0                 8.0    8238.0  
295866                 6.0                 8.0    5278.0  
...                    ...                 ...       ...  
231838                15.0                 8.0    8108.0  
231839                 2.0                15.0   15180.0  
231840                 2.0                 8.0   19173.0  
231942                13.0                15.0   10080.0  
0                      6.0                14.0    8370.0  

[550068 rows x 12 columns]

### Task 23. Find top 20 most revenue generated customer and their purchased product id

In [175]:
# Lets get top 20 using iloc
Top20 = df1.sort_values(by=['Purchase'],ascending=False).iloc[:20,:]
Top20

User_ID Product_ID Gender    Age  Occupation City_Category  \
370891  1003160  P00052842      M  26-35          17             C   
93016   1002272  P00052842      M  26-35           0             C   
87440   1001474  P00052842      M  26-35           4             A   
503697  1005596  P00117642      M  36-45          12             B   
321782  1001577  P00052842      M    55+           0             C   
349658  1005848  P00119342      M  51-55          20             A   
292083  1003045  P00052842      M  46-50           1             B   
298378  1003947  P00116142      M  26-35           0             C   
437804  1001387  P00086242      F  51-55          13             B   
229329  1005367  P00085342      M  18-25           4             A   
416883  1004117  P00161842      M  18-25           4             B   
7542    1001178  P00116142      M    55+           0             C   
373300  1003511  P00085342      M  51-55           0             C   
33268   1005102  P00052842      M  26-35          12             C   
388010  1005716  P00052842      M   0-17          10             C   
449656  1003301  P00086242      F  26-35           2             B   
366333  1002359  P00085342      M    55+          13             C   
54364   1002274  P00052842      M  18-25           2             B   
56879   1002788  P00085342      M    55+           1             B   
68926   1004520  P00116142      M  26-35           4             C   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
370891                          3               0                  10   
93016                           1               0                  10   
87440                           2               1                  10   
503697                          1               0                  10   
321782                          1               1                  10   
349658                          0               1                  10   
292083                          2               1                  10   
298378                          3               0                  10   
437804                          1               1                  10   
229329                          1               0                  10   
416883                         4+               0                  10   
7542                            1               1                  10   
373300                          2               1                  10   
33268                           2               0                  10   
388010                         4+               0                  10   
449656                          3               0                  10   
366333                          1               1                  10   
54364                           3               0                  10   
56879                           0               1                  10   
68926                           1               1                  10   

        Product_Category_2  Product_Category_3  Purchase  
370891                15.0                 8.0   23961.0  
93016                 15.0                 8.0   23961.0  
87440                 15.0                 8.0   23961.0  
503697                16.0                 9.0   23960.0  
321782                15.0                17.0   23960.0  
349658                13.0                 5.0   23960.0  
292083                15.0                17.0   23960.0  
298378                13.0                16.0   23959.0  
437804                14.0                 5.0   23959.0  
229329                13.0                16.0   23958.0  
416883                13.0                16.0   23958.0  
7542                  13.0                16.0   23958.0  
373300                13.0                15.0   23958.0  
33268                 15.0                 8.0   23956.0  
388010                15.0                 4.0   23955.0  
449656                 6.0                13.0   23955.0  


In [176]:
# Get list of top 20 user id
Top20.User_ID

370891    1003160
93016     1002272
87440     1001474
503697    1005596
321782    1001577
349658    1005848
292083    1003045
298378    1003947
437804    1001387
229329    1005367
416883    1004117
7542      1001178
373300    1003511
33268     1005102
388010    1005716
449656    1003301
366333    1002359
54364     1002274
56879     1002788
68926     1004520
Name: User_ID, dtype: int64

In [177]:
# Visualize products included in top 20
Top20.Product_ID.value_counts()

P00052842    8
P00085342    4
P00116142    3
P00086242    2
P00117642    1
P00119342    1
P00161842    1
Name: Product_ID, dtype: int64

### Task 24. Find which age group is much active for purchasing product from website

In [178]:
# Lets use unique to get distinct values
df1['Gender'].unique()

array(['F', 'M'], dtype=object)

In [179]:
# Lets use value_counts to get count of distinct values
df1['Gender'].value_counts()

M    414259
F    135809
Name: Gender, dtype: int64

In [180]:
# Lets get age count sorted in descending order
df1['Age'].value_counts(ascending=False)

26-35    219587
36-45    110013
18-25     99660
46-50     45701
51-55     38501
55+       21504
0-17      15102
Name: Age, dtype: int64

In [181]:
# We can also replace for column values
df1['Gender']=df1['Gender'].replace('F','Female')
df1['Gender']=df1['Gender'].replace('M','Male')
#df1['Gender']
df1.head(10)

User_ID Product_ID  Gender    Age  Occupation City_Category  \
0  1000001  P00069042  Female   0-17          10             A   
1  1000001  P00248942  Female   0-17          10             A   
2  1000001  P00087842  Female   0-17          10             A   
3  1000001  P00085442  Female   0-17          10             A   
4  1000002  P00285442    Male    55+          16             C   
5  1000003  P00193542    Male  26-35          15             A   
6  1000004  P00184942    Male  46-50           7             B   
7  1000004  P00346142    Male  46-50           7             B   
8  1000004   P0097242    Male  46-50           7             B   
9  1000005  P00274942    Male  26-35          20             A   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   
5                          3               0                   1   
6                          2               1                   1   
7                          2               1                   1   
8                          2               1                   1   
9                          1               1                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 6.0                14.0    8370.0  
1                 6.0                14.0   15200.0  
2                 6.0                14.0    1422.0  
3                14.0                14.0    1057.0  
4                14.0                14.0    7969.0  
5                 2.0                14.0   15227.0  
6                 8.0                17.0   19215.0  
7                15.0                17.0   15854.0  
8                16.0                17.0   15686.0  
9                16.0                17.0    7871.0

### Task 25. Generate list of User ID with corresponding age and find the total count of purchases thet have done.

In [182]:
counts = df1[['User_ID','Age']].value_counts()
counts

User_ID  Age  
1001680  26-35    1026
1004277  36-45     979
1001941  36-45     898
1001181  36-45     862
1000889  46-50     823
                  ... 
1002111  55+         7
1005391  26-35       7
1002690  26-35       7
1005608  18-25       7
1000708  26-35       6
Length: 5891, dtype: int64

### Task 26. Get different statistical values for Purchase column

### Aggregations in pandas

In [183]:
# Import numpy
import numpy as np

In [184]:
# Lets use describe on Purchase
df1['Purchase'].describe()

count       550068.0
mean     9263.968713
std      5023.065394
min             12.0
25%           5823.0
50%           8047.0
75%          12054.0
max          23961.0
Name: Purchase, dtype: Float64

### Task 27. Find the total amount generated via website by selling product

In [185]:
# Lets use np.sum aggregation to get total puchased amount
df1['Purchase'].aggregate(np.sum)

5095812742.0

In [186]:
# find sum and mean value after doing aggregate over purchase amount
# we use mean to find average
df1['Purchase'].aggregate([np.sum,np.mean])

sum     5.095813e+09
mean    9.263969e+03
Name: Purchase, dtype: float64

In [187]:
# Find mean and sum for 'Product_Category_1', 'Product_Category_2', 'Product_Category_3'
df1[['Product_Category_1', 'Product_Category_2', 'Product_Category_3']].aggregate([np.sum,np.mean])

Product_Category_1  Product_Category_2  Product_Category_3
sum         2.972716e+06        5.425425e+06        6.958758e+06
mean        5.404270e+00        9.863190e+00        1.265072e+01

### Task 28. Tag records to 'high focused' transaction where purchase amount has been more than 5000. Remaining can be tagged as general transaction.

## <u>Function application in pandas</u>

In [188]:
# Try to use apply function on Product_Category_1
df1.Product_Category_1.apply(lambda x : x*10)

0          30
1          10
2         120
3         120
4          80
         ... 
550063    200
550064    200
550065    200
550066    200
550067    200
Name: Product_Category_1, Length: 550068, dtype: int64

In [189]:
# Lets add new column as 'category' which will have tags based on purchase amount
df1['category']=df1.Purchase.apply(lambda x : 'High Focused' if x > 5000 else 'General')
df1

User_ID Product_ID  Gender    Age  Occupation City_Category  \
0       1000001  P00069042  Female   0-17          10             A   
1       1000001  P00248942  Female   0-17          10             A   
2       1000001  P00087842  Female   0-17          10             A   
3       1000001  P00085442  Female   0-17          10             A   
4       1000002  P00285442    Male    55+          16             C   
...         ...        ...     ...    ...         ...           ...   
550063  1006033  P00372445    Male  51-55          13             B   
550064  1006035  P00375436  Female  26-35           1             C   
550065  1006036  P00375436  Female  26-35          15             B   
550066  1006038  P00375436  Female    55+           1             C   
550067  1006039  P00371644  Female  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
550063                          1               1                  20   
550064                          3               0                  20   
550065                         4+               1                  20   
550066                          2               0                  20   
550067                         4+               1                  20   

        Product_Category_2  Product_Category_3  Purchase      category  
0                      6.0                14.0    8370.0  High Focused  
1                      6.0                14.0   15200.0  High Focused  
2                      6.0                14.0    1422.0       General  
3                     14.0                14.0    1057.0       General  
4                     14.0                14.0    7969.0  High Focused  
...                    ...                 ...       ...           ...  
550063                 2.0                11.0     368.0       General  
550064                 2.0                11.0     371.0       General  
550065                 2.0                11.0     137.0       General  
550066                 2.0                11.0     365.0       General  
550067                 2.0                11.0     490.0       General  

[550068 rows x 13 columns]

In [190]:
# Lets check value for high focused row
df1.category.value_counts() #category column counts

High Focused    455145
General          94923
Name: category, dtype: int64

### Task 29. Based on gender, check the total purchased amount and average purchasing amount

## Pandas GroupBy operations

In [191]:
# Use group by on top of gender column
df1.groupby('Gender')

In [192]:
df1.groupby('Gender').groups

{'Female': [0, 1, 2, 3, 14, 15, 16, 17, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 65, 66, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 124, 125, 126, 147, 148, 149, 150, 151, 156, 157, 158, 163, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 219, 222, 223, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 297, 298, 299, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 373, ...], 'Male': [4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 67, 68, 69, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 152, 153, ...]}

## Group by with multiple columns

In [193]:
# Get groups based on gender and age combination
df1.groupby(['Gender','Age']).groups

{('Female', '0-17'): [0, 1, 2, 3, 299, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 467, 468, 539, 540, 541, 542, 543, 617, 618, 619, 620, 621, 1150, 1151, 1304, 1305, 1306, 2905, 2907, 3010, 3715, 3804, 3805, 3806, 3807, 3808, 3835, 3836, 4551, 4552, 4553, 4554, 4555, 5453, 6431, 6759, 6760, 6761, 6762, 6763, 6764, 6765, 6766, 6767, 6768, 6769, 6770, 6771, 6772, 6773, 6774, 6775, 6776, 6777, 6778, 6779, 6780, 6781, 6782, 6783, 6784, 6785, 6786, 6787, 6788, 6789, 6790, 6791, 6792, 6793, 6794, 6795, 6796, 6797, 6798, 6799, 6800, 6801, 6802, 6803, 6804, 6805, 6806, 6807, 6808, ...], ('Female', '18-25'): [70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 222, 223, 495, 496, 497, 498, 499, 500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 547, 548, 549, 550, 625, 910, 911, 912, 913, 914, 1046, 1228, 1267, 1268, 1269, 1490, 1491, 1492, 1493, 1494, 1495, 1496, 1497, 1498, 1499, 1500, 1552, 1553, 1554, 1555, 1556, 1665, 1666, 16

In [194]:
# apply aggregation function sum with groupby
df1.groupby('Gender').sum()

User_ID  Occupation  Marital_Status  Product_Category_1  \
Gender                                                                 
Female  136234060927      915426           56988              776517   
Male    415500008355     3527312          168349             2196199   

        Product_Category_2  Product_Category_3      Purchase  
Gender                                                        
Female           1356094.0           1701913.0  1186232642.0  
Male             4069331.0           5256845.0  3909580100.0

In [195]:
# use np.sum
df1.groupby('Gender').agg(np.sum)

User_ID  Occupation  Marital_Status  Product_Category_1  \
Gender                                                                 
Female  136234060927      915426           56988              776517   
Male    415500008355     3527312          168349             2196199   

        Product_Category_2  Product_Category_3      Purchase  
Gender                                                        
Female           1356094.0           1701913.0  1186232642.0  
Male             4069331.0           5256845.0  3909580100.0

In [196]:
# Get total purchased amount
df1.groupby('Gender')['Purchase'].agg(np.sum)


Gender
Female    1186232642.0
Male      3909580100.0
Name: Purchase, dtype: Float64

In [198]:
# Get sum as well as mean 
df1.groupby('Gender')['Purchase'].agg([np.sum,np.mean])

sum         mean
Gender                           
Female  1186232642.0  8734.565765
Male    3909580100.0   9437.52604

**Glimpse of important activities we did using pandas**

- Generated descriptive statistical values for numerical columns. Also explored categorical columns.
- Got percentage distribution of each product id available in dataset and found with highest occuring product id in dataset.
- Handled missing value using dropping and imputing both options.
- Fetched maximum purchase amount with complete row details.
- Found the purchase amount for a specified user_id and product_id.
- Found some interesting subset of dataset based on given conditions.
- Found top 20 most revenue generated customer and their purchased product id.
- Found which age group is much active for purchasing product from website.
- Found the total amount generated via website by selling product.
- Tagged records to 'high focused' transaction where purchase amount has been more than 5000.
- Based on gender, got the total purchased amount and average purchasing amount.

Thanks :)

  ~ Sejal Sahu